### PR TITLE
Purge unnecessary `DataManager` checks

### LINKED ISSUE #401: Remove unnecessary checks in `DataManager`
The `DataManager` class's methods contain a lot of code that looks like this:

```c++
std::vector<PopulationItem>
DataManager::get_population(const Country &country,
                            std::function<bool(unsigned int)> time_filter) const {
    auto results = std::vector<PopulationItem>();

    if (index_.contains("demographic")) {
        // [snip]
        } else {
            notify_warning(
                fmt::format("{} population file: '{}' not found.", country.name, filename));
        }
    } else {
        notify_warning("index has no 'demographic' entry.");
    }

    return results;
}
```

Now that we're checking the input file with a schema, we know that the index file will contain a `demographic` property, so this check is unnecessary. The only thing is that the current code would seem to suggest that this property is optional rather than required, though really I think it *should* be required. And we should def be throwing an error if the file can't be read!

@jamesturner246 Are you ok with me removing all these unnecessary checks?

### PR DESCRIPTION
**Pull request by alexdewar**
Currently, many of the methods in `DataManager` check whether a certain field is present and, if it isn't, return an empty data structure. Those checks are now unnecessary because we enforce the presence of all required fields with a JSON schema, so they can be removed. In turn, this means some of the callers of these methods no longer need to check the return value.

Additionally, many methods return an empty data structure if a given CSV file is not found. In most cases, I've changed things to throw an exception if the file isn't present, because the files are in the `data/` folder, so this problem shouldn't arise. The one exception is `get_relative_risk_to_risk_factor`, where we actually don't have CSV files for all possible combinations of inputs (HealthGPS will tell you which ones are missing, but you have to run it with the `--verbose` flag to see). In that case, I changed the return type to be `std::optional<RelativeRiskEntity>`, with `std::nullopt` indicating that a file is missing (which is a safer approach).

Closes #401.

### CODE REVIEW

File: src/HealthGPS.Core/datastore.h [hunk 1]
```diff
@@ -49,35 +51,35 @@ class Datastore {
...
+    /// @return The disease definition
     virtual DiseaseEntity get_disease(const DiseaseInfo &info, const Country &country) const = 0;
 
     /// @brief Gets the relative risk effects for disease to disease interactions
     /// @param source The source disease information
     /// @param target The target disease information
-    /// @return The disease-disease effects, check empty() = true for missing data.
+    /// @return The disease-disease effects
     virtual RelativeRiskEntity get_relative_risk_to_disease(const DiseaseInfo &source,
                                                             const DiseaseInfo &target) const = 0;
```
Comment (jamesturner246) [inline src/HealthGPS.Core/datastore.h:62]: It should also be possible to apply the `optional` approach from `get_relative_risk_to_risk_factor` here.

File: src/HealthGPS.Datastore/datamanager.h [hunk 2]
```diff
@@ -58,7 +58,7 @@ class DataManager : public Datastore {
     RelativeRiskEntity get_relative_risk_to_disease(const DiseaseInfo &source,
```
Comment (jamesturner246) [inline src/HealthGPS.Datastore/datamanager.h:58]: See above.

File: src/HealthGPS/analysis_module.cpp [hunk 3]
```diff
@@ -400,7 +400,7 @@ std::unique_ptr<AnalysisModule> build_analysis_module(Repository &repository,
                                                       const ModelInput &config) {
     auto analysis_entity = repository.manager().get_disease_analysis(config.settings().country());
     auto &lms_definition = repository.get_lms_definition();
-    if (analysis_entity.empty() || lms_definition.empty()) {
+    if (lms_definition.empty()) {
         throw std::logic_error(
             "Failed to create analysis module, invalid disease analysis definition.");
```
Comment (jamesturner246) [inline src/HealthGPS/analysis_module.cpp:405]:
```suggestion
            "Failed to create analysis module: invalid LMS model definition.");
```

File: src/HealthGPS/converter.cpp [hunk 4]
```diff
@@ -78,26 +78,22 @@ RelativeRiskLookup StoreConverter::to_relative_risk_lookup(const core::RelativeR
 RelativeRisk create_relative_risk(const RelativeRiskInfo &info) {
     RelativeRisk result;
     for (const auto &item : info.inputs.diseases()) {
-        auto table = info.manager.get_relative_risk_to_disease(info.disease, item);
-        if (!table.empty() && !table.is_default_value) {
+        const auto table = info.manager.get_relative_risk_to_disease(info.disease, item);
+        if (!table.is_default_value) {
```
Comment (jamesturner246) [inline src/HealthGPS/converter.cpp:82]: Is the `!table.is_default_value` check still necessary if we aren't doing default tables anymore?

**Review by jamesturner246 — CHANGES_REQUESTED** [review]
Looks good, but I think the `relative_risk_from_disease` can get the same treatment. Also a couple of questions.

File: src/HealthGPS.Datastore/datamanager.cpp [hunk 5]
```diff
@@ -166,243 +158,213 @@ std::vector<MortalityItem>
...
+        auto row_data = std::vector<float>(row_size);
+        for (size_t col = 0; col < row_size; col++) {
+            row_data[col] = std::stof(row[col]);
         }
-        notify_warning(fmt::format("{} to {} relative risk file not found, using default.",
-                                   source_code_str, target_code_str));
 
-        return generate_default_relative_risk_to_disease();
+        non_default_count +=
+            std::count_if(std::next(row_data.cbegin()), row_data.cend(), [&default_value](auto &v) {
```
[the remark this answers is not in the document]
Comment (alexdewar) [reply src/HealthGPS.Datastore/datamanager.cpp:349]: Interesting... this looks like a new `clang-tidy` check. I ran `clang-tidy` the other day and there were *loads* of these warnings, in basically every file.

I think we should suppress this one. I'll open a new PR for it.

[the remark this answers is not in the document]
Comment (alexdewar) [reply src/HealthGPS.Datastore/datamanager.cpp:349]: This is just silly. Another check for the chopping block!

File: src/HealthGPS.Datastore/datamanager.cpp [hunk 6]
```diff
@@ -522,60 +477,50 @@ std::vector<BirthItem> DataManager::get_birth_indicators(const Country &country)
...
+    filename = (root_ / nodepath / filefolder / filename).string();
+    if (!std::filesystem::exists(filename)) {
+        throw std::runtime_error(fmt::format("{}, demographic indicators file: '{}' not found.",
+                                             country.name, filename));
+    }
 
-            result.push_back(BirthItem{.at_time = row_time,
-                                       .number = std::stof(row[mapping["Births"]]),
-                                       .sex_ratio = std::stof(row[mapping["SRB"]])});
+    rapidcsv::Document doc(filename);
```
[the remark this answers is not in the document]
Comment (alexdewar) [reply src/HealthGPS.Datastore/datamanager.cpp:493]: I'm actually on board with this one, but what kind of a psychopath puts the `const` *after* the type?!

File: src/HealthGPS/converter.cpp (hunk 4 above)
Comment (alexdewar) [reply src/HealthGPS/converter.cpp:82]: Indeed it's not. Let's just return a `std::optional` instead and drop this field

File: src/HealthGPS.Datastore/datamanager.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (alexdewar) [reply src/HealthGPS.Datastore/datamanager.cpp:350]: Weirdly, this check should already be being suppressed: https://github.com/imperialCHEPI/healthgps/blob/993e8b112e6c1a7109ba28446bb9d84fffc82e94/.clang-tidy#L22

File: src/HealthGPS.Datastore/datamanager.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (alexdewar) [reply src/HealthGPS.Datastore/datamanager.cpp:493]: This one should be ignored too: https://github.com/imperialCHEPI/healthgps/blob/993e8b112e6c1a7109ba28446bb9d84fffc82e94/.clang-tidy#L27

File: src/HealthGPS.Tests/Datastore.Test.cpp [hunk 7]
```diff
@@ -192,31 +186,23 @@ TEST_F(DatastoreTest, DiseaseRelativeRiskToDisease) {
...
-    ASSERT_TRUE(table_self.is_default_value);
-
-    ASSERT_EQ(3, table_other.columns.size());
-    ASSERT_GT(table_other.rows.size(), 0);
-    ASSERT_EQ(table_other.rows[0][1], table_other.rows[0][2]);
-    ASSERT_FALSE(table_other.is_default_value);
+    EXPECT_FALSE(table_self.has_value());
+
+    ASSERT_TRUE(table_other.has_value());
+    EXPECT_EQ(3, table_other->columns.size());
```
[the remark this answers is not in the document]
Comment (alexdewar) [reply src/HealthGPS.Tests/Datastore.Test.cpp:192]: It's actually checked by the assert statement on the previous line, but clearly `clang-tidy` can't tell. I guess we have to go back to suppressing the warning manually...

**Review by jamesturner246 — APPROVED** [review]
LGTM :+1:

**Pull request merged**